### PR TITLE
fix(ui): fix no-restrict-syntax in ui

### PR DESCRIPTION
**Pull request by simeng-li**
## Summary

Fix most of the no-restricted-syntax in UI. 
Disable the lint for some corner cases. 

Introduce superstruct to the UI package. Use superstruct to guard some unknown input type like location.state. 

## Testing

test locally
@logto-io/eng

### CODE REVIEW

File: packages/ui/src/hooks/use-bind-social.ts [hunk 1]
```diff
@@ -54,7 +51,7 @@ const useBindSocial = () => {
 
   return {
     localSignInMethods,
-    relatedUser: state?.relatedUser,
+    relatedUser: is(state, bindSocialStateGuard) ? state.relatedUser : undefined,
```
Comment (gao-sun) [inline packages/ui/src/hooks/use-bind-social.ts:54]: conditional?

File: packages/ui/src/hooks/use-form.ts [hunk 2]
```diff
@@ -24,6 +24,7 @@ const useForm = <T>(initialState: T) => {
       fieldValidationsRef.current[key]?.(value),
     ]);
 
+    // eslint-disable-next-line no-restricted-syntax
```
Comment (gao-sun) [inline packages/ui/src/hooks/use-form.ts:27]: better to leave the reason here?

File: packages/ui/src/hooks/use-form.ts [hunk 3]
```diff
@@ -37,8 +38,8 @@ const useForm = <T>(initialState: T) => {
       return {
         value: fieldValue[field],
         error: fieldErrors[field],
-        onChange: ({ target }: FormEvent<HTMLInputElement>) => {
-          const { value } = target as HTMLInputElement;
+        onChange: ({ currentTarget }: FormEvent<HTMLInputElement>) => {
+          const { value } = currentTarget;
```
Comment (gao-sun) [inline packages/ui/src/hooks/use-form.ts:42]:
```suggestion
        onChange: ({ currentTarget: { value } }: FormEvent<HTMLInputElement>) => {
```

File: packages/ui/src/hooks/use-form.ts [hunk 4]
```diff
@@ -55,6 +56,7 @@ const useForm = <T>(initialState: T) => {
         previous[key] && fieldValidationsRef.current[key]?.(value),
       ]);
 
+      // eslint-disable-next-line no-restricted-syntax
```
Comment (gao-sun) [inline packages/ui/src/hooks/use-form.ts:59]: ditto

File: packages/ui/src/hooks/use-preview.ts [hunk 5]
```diff
@@ -44,6 +35,7 @@ const usePreview = (context: Context): [boolean, PreviewConfig?] => {
       }
 
       if (event.data.sender === 'ac_preview') {
+        // eslint-disable-next-line no-restricted-syntax
```
Comment (gao-sun) [inline packages/ui/src/hooks/use-preview.ts:38]: ditto

File: packages/ui/src/utils/index.ts [hunk 6]
```diff
@@ -26,16 +26,16 @@ export const getSearchParameters = (parameters: string | URLSearchParams, key: s
...
 };
 
-type Entries<T> = Array<
+export type Entries<T> = Array<
   {
     [K in keyof T]: [K, T[K]];
   }[keyof T]
 >;
 
+// eslint-disable-next-line no-restricted-syntax
```
Comment (gao-sun) [inline packages/ui/src/utils/index.ts:35]: or we can have a `fromEntries()` as well if the type issue cannot fix

File: packages/ui/src/hooks/use-bind-social.ts (hunk 1 above)
Comment (simeng-li) [reply packages/ui/src/hooks/use-bind-social.ts]: updated

File: packages/ui/src/hooks/use-form.ts (hunk 2 above)
Comment (simeng-li) [reply packages/ui/src/hooks/use-form.ts]: updated extract as a util fromEntries

File: packages/ui/src/hooks/use-form.ts (hunk 3 above)
Comment (simeng-li) [reply packages/ui/src/hooks/use-form.ts]: updated

File: packages/ui/src/utils/index.ts (hunk 6 above)
Comment (simeng-li) [reply packages/ui/src/utils/index.ts:35]: updated

File: packages/ui/src/hooks/use-preview.ts (hunk 5 above)
Comment (simeng-li) [reply packages/ui/src/hooks/use-preview.ts:39]: comments added

File: packages/ui/src/utils/index.ts [hunk 7]
```diff
@@ -26,16 +26,21 @@ export const getSearchParameters = (parameters: string | URLSearchParams, key: s
...
 
-// eslint-disable-next-line @typescript-eslint/ban-types
-export const inOperator = <K extends string, T extends object>(
-  key: K,
+export const fromEntries = <T extends Record<string, unknown>>(entries: Entries<T>) =>
+  // eslint-disable-next-line no-restricted-syntax
+  Object.fromEntries(entries) as T;
+
+export const isKeyOf = <T extends Record<string, unknown>>(
+  key: string | number | symbol,
```
Comment (gao-sun) [inline packages/ui/src/utils/index.ts:44]:
```suggestion
  key: string,
```

**Review by gao-sun — APPROVED** [review]
